### PR TITLE
Support ops.sum(data, dim=None, keepdims=False)

### PR DESCRIPTION
**Pull request by fritzo**
Addresses #489 
pair coded with @eb8680 @ordabayevy @fehiepsi 

This demonstrates the new parametrized op syntax from #491 .  The recipe is:
1. add new `*args, *kwargs` to your op in `funsor.ops.array`
2. add backend support in `funsor.torch.ops` and `funsor.jax.ops`
3. implement `find_domain(op, ...)` in `funsor.domains`
4. work around batch dimensions in `funsor.tensor`
5. add some tests

## Tested
- [x] unit test for various op arg,kwarg syntax
- [x] unit test for batched tensors

### CODE REVIEW

**Comment by fritzo** [comment]
@ordabayevy I hope this can serve as a template for your #482 .  Feel free to refactor after this PR merges, in case you'd like to reuse the logic e.g. in `find_domain` or `eager.register`. We may even be able to create a subclass `ReductionOp(UnaryOp)` and register `ops.sum`, `ops.mean`, etc. in a single pattern.

File: funsor/tensor.py [hunk 1]
```diff
@@ -772,6 +772,25 @@ def eager_reshape_tensor(op, arg):
     return Tensor(data, arg.inputs, arg.dtype)
 
 
+@eager.register(Unary, ops.SumOp, Tensor)
+def eager_reshape_tensor(op, arg):
+    if not arg.inputs:
+        return Tensor(op(arg.data), arg.inputs, arg.dtype)
+
+    # Work around batch inputs.
```
Comment (eb8680) [inline funsor/tensor.py:780]: This logic converting dimensions into batch-aware dimensions seems useful and general-purpose enough that some version of it should maybe live in an `Op` method or something - we don't want to have to write this from scratch in each new op with nontrivial shape semantics.

File: funsor/tensor.py [hunk 2]
```diff
@@ -772,6 +772,25 @@ def eager_reshape_tensor(op, arg):
     return Tensor(data, arg.inputs, arg.dtype)
 
 
+@eager.register(Unary, ops.SumOp, Tensor)
+def eager_reshape_tensor(op, arg):
```
Comment (eb8680) [inline funsor/tensor.py:776]: nit: `eager_sum_tensor`

File: funsor/tensor.py (hunk 1 above)
Comment (fritzo) [reply funsor/tensor.py:780]: I agree this is useful, and I'm hoping @ordabayevy will generalize this in #482 once we have more than one use case.  This file seems like the right place for that general logic since `funsor.ops` should be agnostic to `inputs` and `domains` etc.

**Review by eb8680 — APPROVED** [review]
